// @ladderlabs/buddy-sdk/dist/cjs/client/providers/AmbassadorProvider.js
"use strict";
exports.__esModule = true;
exports.AmbassadorProvider = void 0;
var tslib_1 = require("tslib");
var pda_1 = require("../../utils/pda");
var Ambassador_1 = require("../../models/Ambassador");
var type_1 = require("../../utils/type");
var AmbassadorProvider = /** @class */ (function () {
    function AmbassadorProvider(program, wallet) {
        this.program = program;
        this.wallet = wallet;
    }
    AmbassadorProvider.prototype.getAll = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var ambassadors;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.ambassador.all()];
                    case 1:
                        ambassadors = _a.sent();
                        return [2 /*return*/, ambassadors.map(function (ambassador) {
                                return new Ambassador_1.Ambassador(tslib_1.__assign({ pda: ambassador.publicKey }, ambassador.account), _this.program);
                            })];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByWallet = function (wallet) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var ambassadors;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.wallet && wallet)
                            throw type_1.NO_WALLET_ERROR;
                        return [4 /*yield*/, this.program.account.ambassador.all([
                                {
                                    memcmp: {
                                        offset: 8,
                                        bytes: wallet ? wallet.toBase58() : this.wallet.toBase58()
                                    }
                                }
                            ])];
                    case 1:
                        ambassadors = _a.sent();
                        return [2 /*return*/, ambassadors.map(function (ambassador) {
                                return new Ambassador_1.Ambassador(tslib_1.__assign({ pda: ambassador.publicKey }, ambassador.account), _this.program);
                            })];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByOrganization = function (organization) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getByPDA(organization.account.ambassador)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByName = function (name) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var ambassadorPDA;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ambassadorPDA = (0, pda_1.getAmbassadorPDA)(this.program, name.toLowerCase());
                        return [4 /*yield*/, this.getByPDA(ambassadorPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByPDA = function (ambassadorPDA) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var ambassador;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.ambassador.fetchNullable(ambassadorPDA)];
                    case 1:
                        ambassador = _a.sent();
                        if (!ambassador)
                            return [2 /*return*/, ambassador];
                        return [2 /*return*/, new Ambassador_1.Ambassador(tslib_1.__assign(tslib_1.__assign({}, ambassador), { pda: ambassadorPDA }), this.program, this.wallet)];
                }
            });
        });
    };
    return AmbassadorProvider;
}());
exports.AmbassadorProvider = AmbassadorProvider;
