// @ladderlabs/buddy-sdk/dist/cjs/client/providers/MemberProvider.js
"use strict";
exports.__esModule = true;
exports.MemberProvider = void 0;
var tslib_1 = require("tslib");
var Member_1 = require("../../models/Member");
var pda_1 = require("../../utils/pda");
var type_1 = require("../../utils/type");
var utils_1 = require("../builders/utils/utils");
var bytes_1 = require("@project-serum/anchor/dist/cjs/utils/bytes");
var MemberProvider = /** @class */ (function () {
    function MemberProvider(program, wallet) {
        this.program = program;
        this.wallet = wallet;
    }
    MemberProvider.prototype.isMemberAvailable = function (organizationName, memberName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var memberPDA, buddyNamePDA;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        memberPDA = (0, pda_1.getMemberPDA)(this.program, organizationName.toString(), memberName.toLowerCase());
                        buddyNamePDA = (0, pda_1.getBuddyNamePDA)(this.program, memberName);
                        return [4 /*yield*/, this.program.account.nameCheck.fetchNullable(buddyNamePDA)];
                    case 1:
                        if (_a.sent()) {
                            return [2 /*return*/, false];
                        }
                        return [4 /*yield*/, this.program.account.member.fetchNullable(memberPDA)];
                    case 2: return [2 /*return*/, !(_a.sent())];
                }
            });
        });
    };
    MemberProvider.prototype.getAll = function (organizationName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!organizationName) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.program.account.member.all([{
                                    memcmp: {
                                        offset: 154,
                                        bytes: bytes_1.bs58.encode(Buffer.from(organizationName))
                                    }
                                }])];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2: return [4 /*yield*/, this.program.account.member.all()];
                    case 3: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    MemberProvider.prototype.getAllStatistics = function () {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.memberStatistics.all()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    MemberProvider.prototype.getByWallet = function (organizationName, wallet) {
        var _a;
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var organization, profile, treasuryPDA, treasury, members;
            var _this = this;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.wallet && wallet)
                            throw type_1.NO_WALLET_ERROR;
                        return [4 /*yield*/, (0, utils_1.getOrganization)(this.program, organizationName)];
                    case 1:
                        organization = _b.sent();
                        return [4 /*yield*/, this.program.account.buddy.all([
                                {
                                    memcmp: {
                                        offset: 8,
                                        bytes: wallet ? wallet.toBase58() : (_a = this.wallet) === null || _a === void 0 ? void 0 : _a.toBase58()
                                    }
                                },
                            ])];
                    case 2:
                        profile = (_b.sent())[0];
                        if (!profile)
                            return [2 /*return*/, []];
                        treasuryPDA = (0, pda_1.getTreasuryPDA)(this.program, [profile.publicKey], [10000], organization.mainTokenMint);
                        return [4 /*yield*/, this.program.account.treasury.fetchNullable(treasuryPDA)];
                    case 3:
                        treasury = _b.sent();
                        if (!treasury)
                            return [2 /*return*/, []];
                        return [4 /*yield*/, this.program.account.member.all([
                                { memcmp: { offset: 41, bytes: treasuryPDA.toBase58() } },
                            ])];
                    case 4:
                        members = _b.sent();
                        return [2 /*return*/, members.map(function (member) { return new Member_1.Member(tslib_1.__assign({ pda: member.publicKey }, member.account), _this.program); })];
                }
            });
        });
    };
    MemberProvider.prototype.getByPDA = function (memberPDA) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var member;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.member.fetchNullable(memberPDA)];
                    case 1:
                        member = _a.sent();
                        if (!member)
                            return [2 /*return*/, null];
                        return [2 /*return*/, new Member_1.Member(tslib_1.__assign(tslib_1.__assign({}, member), { pda: memberPDA }), this.program, this.wallet)];
                }
            });
        });
    };
    MemberProvider.prototype.getByName = function (organizationName, memberName) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var memberPDA;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        memberPDA = (0, pda_1.getMemberPDA)(this.program, organizationName.toLowerCase(), memberName.toLowerCase());
                        return [4 /*yield*/, this.getByPDA(memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    MemberProvider.prototype.getByTreasuryOwner = function (treasuryPDA) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var members;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.member.all([
                            { memcmp: { offset: 41, bytes: treasuryPDA.toBase58() } },
                        ])];
                    case 1:
                        members = (_a.sent()).map(function (member) {
                            return new Member_1.Member(tslib_1.__assign(tslib_1.__assign({}, member.account), { pda: member.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, members];
                }
            });
        });
    };
    MemberProvider.prototype.getByTreasuryReferrer = function (treasuryPDA) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var members;
            var _this = this;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.member.all([
                            { memcmp: { offset: 9, bytes: treasuryPDA.toBase58() } },
                        ])];
                    case 1:
                        members = (_a.sent()).map(function (member) {
                            return new Member_1.Member(tslib_1.__assign(tslib_1.__assign({}, member.account), { pda: member.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, members];
                }
            });
        });
    };
    return MemberProvider;
}());
exports.MemberProvider = MemberProvider;
