// @ladderlabs/buddy-sdk/dist/cjs/client/providers/PDA.js
"use strict";
exports.__esModule = true;
exports.PDA = void 0;
var pda_1 = require("../../utils/pda");
var PDA = /** @class */ (function () {
    function PDA(program) {
        this.program = program;
    }
    PDA.prototype.getMasterTreasuryPDA = function (mint) {
        return (0, pda_1.getMasterTreasuryPDA)(this.program, mint);
    };
    PDA.prototype.getAmbassadorPDA = function (name) {
        return (0, pda_1.getAmbassadorPDA)(this.program, name);
    };
    PDA.prototype.getOrganizationPDA = function (name) {
        return (0, pda_1.getOrganizationPDA)(this.program, name);
    };
    PDA.prototype.getMemberPDA = function (organizationName, name) {
        return (0, pda_1.getMemberPDA)(this.program, organizationName, name);
    };
    PDA.prototype.getMemberStatisticPDA = function (organizationName, memberPDA) {
        return (0, pda_1.getMemberStatisticPDA)(this.program, organizationName, memberPDA);
    };
    PDA.prototype.getBuddyPDA = function (name) {
        return (0, pda_1.getBuddyPDA)(this.program, name);
    };
    PDA.prototype.getProfilePDA = function (name) {
        return (0, pda_1.getBuddyPDA)(this.program, name);
    };
    PDA.prototype.getTreasuryPDA = function (owners, shares, mint) {
        return (0, pda_1.getTreasuryPDA)(this.program, owners, shares, mint);
    };
    PDA.prototype.getSimpleTreasuryPDA = function (buddyPDA, mint) {
        return (0, pda_1.getTreasuryPDA)(this.program, [buddyPDA], [10000], mint);
    };
    return PDA;
}());
exports.PDA = PDA;
