// @ladderlabs/buddy-sdk/dist/esm/client/providers/AmbassadorProvider.js
import { __assign, __awaiter, __generator } from "tslib";
import { getAmbassadorPDA } from "../../utils/pda";
import { Ambassador } from "../../models/Ambassador";
import { NO_WALLET_ERROR } from "../../utils/type";
var AmbassadorProvider = /** @class */ (function () {
    function AmbassadorProvider(program, wallet) {
        this.program = program;
        this.wallet = wallet;
    }
    AmbassadorProvider.prototype.getAll = function () {
        return __awaiter(this, void 0, void 0, function () {
            var ambassadors;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.ambassador.all()];
                    case 1:
                        ambassadors = _a.sent();
                        return [2 /*return*/, ambassadors.map(function (ambassador) {
                                return new Ambassador(__assign({ pda: ambassador.publicKey }, ambassador.account), _this.program);
                            })];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByWallet = function (wallet) {
        return __awaiter(this, void 0, void 0, function () {
            var ambassadors;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.wallet && wallet)
                            throw NO_WALLET_ERROR;
                        return [4 /*yield*/, this.program.account.ambassador.all([
                                {
                                    memcmp: {
                                        offset: 8,
                                        bytes: wallet ? wallet.toBase58() : this.wallet.toBase58()
                                    }
                                }
                            ])];
                    case 1:
                        ambassadors = _a.sent();
                        return [2 /*return*/, ambassadors.map(function (ambassador) {
                                return new Ambassador(__assign({ pda: ambassador.publicKey }, ambassador.account), _this.program);
                            })];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByOrganization = function (organization) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.getByPDA(organization.account.ambassador)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByName = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var ambassadorPDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        ambassadorPDA = getAmbassadorPDA(this.program, name.toLowerCase());
                        return [4 /*yield*/, this.getByPDA(ambassadorPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    AmbassadorProvider.prototype.getByPDA = function (ambassadorPDA) {
        return __awaiter(this, void 0, void 0, function () {
            var ambassador;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.ambassador.fetchNullable(ambassadorPDA)];
                    case 1:
                        ambassador = _a.sent();
                        if (!ambassador)
                            return [2 /*return*/, ambassador];
                        return [2 /*return*/, new Ambassador(__assign(__assign({}, ambassador), { pda: ambassadorPDA }), this.program, this.wallet)];
                }
            });
        });
    };
    return AmbassadorProvider;
}());
export { AmbassadorProvider };
