// @ladderlabs/buddy-sdk/dist/esm/client/providers/BuddyProvider.js
import { __assign, __awaiter, __generator, __spreadArray } from "tslib";
import { Buddy } from "../../models/Buddy";
import { getBuddyPDA } from "../../utils/pda";
import { BUDDY_DOESNT_EXIST, NO_WALLET_ERROR, TREASURY_DOESNT_EXIST } from "../../utils/type";
var BuddyProvider = /** @class */ (function () {
    function BuddyProvider(program, wallet) {
        this.program = program;
        this.wallet = wallet;
    }
    BuddyProvider.prototype.isBuddyAvailable = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var buddyPDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        buddyPDA = getBuddyPDA(this.program, name.toLowerCase());
                        return [4 /*yield*/, this.program.account.buddy.fetchNullable(buddyPDA)];
                    case 1: return [2 /*return*/, !!(_a.sent())];
                }
            });
        });
    };
    BuddyProvider.prototype.getByMember = function (member) {
        return __awaiter(this, void 0, void 0, function () {
            var treasuryPDA, treasury, buddyPDAs, buddies, buddyObjects, i;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        treasuryPDA = member.account.owner;
                        return [4 /*yield*/, this.program.account.treasury.fetchNullable(treasuryPDA)];
                    case 1:
                        treasury = _a.sent();
                        if (!treasury)
                            throw TREASURY_DOESNT_EXIST;
                        buddyPDAs = treasury.owners.map(function (owner) { return owner.ownerPda; });
                        return [4 /*yield*/, this.program.account.buddy.fetchMultiple(buddyPDAs)];
                    case 2:
                        buddies = _a.sent();
                        buddyObjects = [];
                        for (i = 0; i < buddyPDAs.length; i++) {
                            if (buddies[i] === null)
                                throw BUDDY_DOESNT_EXIST;
                            buddyObjects.push(new Buddy(__assign({ pda: buddyPDAs[i] }, (buddies[i])), this.program));
                        }
                        return [2 /*return*/, buddyObjects];
                }
            });
        });
    };
    BuddyProvider.prototype.getByWallet = function (wallet) {
        return __awaiter(this, void 0, void 0, function () {
            var profile, paidBuddies;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!this.wallet && wallet)
                            throw NO_WALLET_ERROR;
                        return [4 /*yield*/, this.getProfile(wallet)];
                    case 1:
                        profile = _a.sent();
                        if (!profile)
                            return [2 /*return*/, []];
                        return [4 /*yield*/, this.getAllByProfile(profile.account.pda)];
                    case 2:
                        paidBuddies = _a.sent();
                        return [2 /*return*/, __spreadArray(__spreadArray([], paidBuddies, true), [profile], false)];
                }
            });
        });
    };
    BuddyProvider.prototype.getAllBuddies = function () {
        return __awaiter(this, void 0, void 0, function () {
            var buddies;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.buddy.all()];
                    case 1:
                        buddies = (_a.sent()).map(function (buddy) {
                            return new Buddy(__assign(__assign({}, buddy.account), { pda: buddy.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, buddies];
                }
            });
        });
    };
    BuddyProvider.prototype.getByPDA = function (buddyPDA) {
        return __awaiter(this, void 0, void 0, function () {
            var buddy;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.buddy.fetchNullable(buddyPDA)];
                    case 1:
                        buddy = _a.sent();
                        if (!buddy)
                            return [2 /*return*/, null];
                        return [2 /*return*/, new Buddy(__assign(__assign({}, buddy), { pda: buddyPDA }), this.program, this.wallet)];
                }
            });
        });
    };
    BuddyProvider.prototype.getByName = function (name) {
        return __awaiter(this, void 0, void 0, function () {
            var buddyPDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        buddyPDA = getBuddyPDA(this.program, name.toLowerCase());
                        return [4 /*yield*/, this.getByPDA(buddyPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    BuddyProvider.prototype.getAllByProfile = function (profilePDA) {
        return __awaiter(this, void 0, void 0, function () {
            var buddies;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.buddy.all([
                            { memcmp: { offset: 8, bytes: profilePDA.toBase58() } },
                        ])];
                    case 1:
                        buddies = (_a.sent()).map(function (buddy) {
                            return new Buddy(__assign(__assign({}, buddy.account), { pda: buddy.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, buddies];
                }
            });
        });
    };
    BuddyProvider.prototype.getProfile = function (wallet) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var buddies;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.wallet && !wallet)
                            throw NO_WALLET_ERROR;
                        return [4 /*yield*/, this.program.account.buddy.all([
                                {
                                    memcmp: {
                                        offset: 8,
                                        bytes: wallet ? wallet.toBase58() : (_a = this.wallet) === null || _a === void 0 ? void 0 : _a.toBase58()
                                    }
                                },
                            ])];
                    case 1:
                        buddies = (_b.sent()).map(function (buddy) {
                            return new Buddy(__assign(__assign({}, buddy.account), { pda: buddy.publicKey }), _this.program, _this.wallet);
                        });
                        //always only have 1 profile
                        return [2 /*return*/, buddies[0] || null];
                }
            });
        });
    };
    return BuddyProvider;
}());
export { BuddyProvider };
