// @ladderlabs/buddy-sdk/dist/esm/client/providers/MemberProvider.js
import { __assign, __awaiter, __generator } from "tslib";
import { Member } from "../../models/Member";
import { getBuddyNamePDA, getMemberPDA, getTreasuryPDA } from "../../utils/pda";
import { NO_WALLET_ERROR } from "../../utils/type";
import { getOrganization } from "../builders/utils/utils";
import { bs58 } from "@project-serum/anchor/dist/cjs/utils/bytes";
var MemberProvider = /** @class */ (function () {
    function MemberProvider(program, wallet) {
        this.program = program;
        this.wallet = wallet;
    }
    MemberProvider.prototype.isMemberAvailable = function (organizationName, memberName) {
        return __awaiter(this, void 0, void 0, function () {
            var memberPDA, buddyNamePDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        memberPDA = getMemberPDA(this.program, organizationName.toString(), memberName.toLowerCase());
                        buddyNamePDA = getBuddyNamePDA(this.program, memberName);
                        return [4 /*yield*/, this.program.account.nameCheck.fetchNullable(buddyNamePDA)];
                    case 1:
                        if (_a.sent()) {
                            return [2 /*return*/, false];
                        }
                        return [4 /*yield*/, this.program.account.member.fetchNullable(memberPDA)];
                    case 2: return [2 /*return*/, !(_a.sent())];
                }
            });
        });
    };
    MemberProvider.prototype.getAll = function (organizationName) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (!organizationName) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.program.account.member.all([{
                                    memcmp: {
                                        offset: 154,
                                        bytes: bs58.encode(Buffer.from(organizationName))
                                    }
                                }])];
                    case 1: return [2 /*return*/, _a.sent()];
                    case 2: return [4 /*yield*/, this.program.account.member.all()];
                    case 3: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    MemberProvider.prototype.getAllStatistics = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.memberStatistics.all()];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    MemberProvider.prototype.getByWallet = function (organizationName, wallet) {
        var _a;
        return __awaiter(this, void 0, void 0, function () {
            var organization, profile, treasuryPDA, treasury, members;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (!this.wallet && wallet)
                            throw NO_WALLET_ERROR;
                        return [4 /*yield*/, getOrganization(this.program, organizationName)];
                    case 1:
                        organization = _b.sent();
                        return [4 /*yield*/, this.program.account.buddy.all([
                                {
                                    memcmp: {
                                        offset: 8,
                                        bytes: wallet ? wallet.toBase58() : (_a = this.wallet) === null || _a === void 0 ? void 0 : _a.toBase58()
                                    }
                                },
                            ])];
                    case 2:
                        profile = (_b.sent())[0];
                        if (!profile)
                            return [2 /*return*/, []];
                        treasuryPDA = getTreasuryPDA(this.program, [profile.publicKey], [10000], organization.mainTokenMint);
                        return [4 /*yield*/, this.program.account.treasury.fetchNullable(treasuryPDA)];
                    case 3:
                        treasury = _b.sent();
                        if (!treasury)
                            return [2 /*return*/, []];
                        return [4 /*yield*/, this.program.account.member.all([
                                { memcmp: { offset: 41, bytes: treasuryPDA.toBase58() } },
                            ])];
                    case 4:
                        members = _b.sent();
                        return [2 /*return*/, members.map(function (member) { return new Member(__assign({ pda: member.publicKey }, member.account), _this.program); })];
                }
            });
        });
    };
    MemberProvider.prototype.getByPDA = function (memberPDA) {
        return __awaiter(this, void 0, void 0, function () {
            var member;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.member.fetchNullable(memberPDA)];
                    case 1:
                        member = _a.sent();
                        if (!member)
                            return [2 /*return*/, null];
                        return [2 /*return*/, new Member(__assign(__assign({}, member), { pda: memberPDA }), this.program, this.wallet)];
                }
            });
        });
    };
    MemberProvider.prototype.getByName = function (organizationName, memberName) {
        return __awaiter(this, void 0, void 0, function () {
            var memberPDA;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        memberPDA = getMemberPDA(this.program, organizationName.toLowerCase(), memberName.toLowerCase());
                        return [4 /*yield*/, this.getByPDA(memberPDA)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    MemberProvider.prototype.getByTreasuryOwner = function (treasuryPDA) {
        return __awaiter(this, void 0, void 0, function () {
            var members;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.member.all([
                            { memcmp: { offset: 41, bytes: treasuryPDA.toBase58() } },
                        ])];
                    case 1:
                        members = (_a.sent()).map(function (member) {
                            return new Member(__assign(__assign({}, member.account), { pda: member.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, members];
                }
            });
        });
    };
    MemberProvider.prototype.getByTreasuryReferrer = function (treasuryPDA) {
        return __awaiter(this, void 0, void 0, function () {
            var members;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.program.account.member.all([
                            { memcmp: { offset: 9, bytes: treasuryPDA.toBase58() } },
                        ])];
                    case 1:
                        members = (_a.sent()).map(function (member) {
                            return new Member(__assign(__assign({}, member.account), { pda: member.publicKey }), _this.program, _this.wallet);
                        });
                        return [2 /*return*/, members];
                }
            });
        });
    };
    return MemberProvider;
}());
export { MemberProvider };
