// @ladderlabs/buddy-sdk/dist/esm/client/providers/PDA.js
import { getAmbassadorPDA, getBuddyPDA, getMasterTreasuryPDA, getMemberPDA, getMemberStatisticPDA, getOrganizationPDA, getTreasuryPDA } from "../../utils/pda";
var PDA = /** @class */ (function () {
    function PDA(program) {
        this.program = program;
    }
    PDA.prototype.getMasterTreasuryPDA = function (mint) {
        return getMasterTreasuryPDA(this.program, mint);
    };
    PDA.prototype.getAmbassadorPDA = function (name) {
        return getAmbassadorPDA(this.program, name);
    };
    PDA.prototype.getOrganizationPDA = function (name) {
        return getOrganizationPDA(this.program, name);
    };
    PDA.prototype.getMemberPDA = function (organizationName, name) {
        return getMemberPDA(this.program, organizationName, name);
    };
    PDA.prototype.getMemberStatisticPDA = function (organizationName, memberPDA) {
        return getMemberStatisticPDA(this.program, organizationName, memberPDA);
    };
    PDA.prototype.getBuddyPDA = function (name) {
        return getBuddyPDA(this.program, name);
    };
    PDA.prototype.getProfilePDA = function (name) {
        return getBuddyPDA(this.program, name);
    };
    PDA.prototype.getTreasuryPDA = function (owners, shares, mint) {
        return getTreasuryPDA(this.program, owners, shares, mint);
    };
    PDA.prototype.getSimpleTreasuryPDA = function (buddyPDA, mint) {
        return getTreasuryPDA(this.program, [buddyPDA], [10000], mint);
    };
    return PDA;
}());
export { PDA };
