// @ladderlabs/buddy-sdk/dist/cjs/instructions/transfers/transferUp.js
"use strict";
exports.__esModule = true;
exports.transferUp = void 0;
var tslib_1 = require("tslib");
var spl_token_1 = require("@solana/spl-token");
var anchor = tslib_1.__importStar(require("@project-serum/anchor"));
var pda_1 = require("../../utils/pda");
var pda_2 = require("../../utils/pda");
var type_1 = require("../../utils/type");
var parse_1 = require("../../utils/parse");
var initTreasury_1 = require("../init/initTreasury");
function transferUp(program, authority, buddyPDA, treasuryPDA) {
    var _a;
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var instructions, treasury, masterOrg, masterTreasuryPDA, masterTreasuryATA, buddy, profilePDA, profile, buddyTreasuryATA, referrerTreasury, referrerTreasuryRewards, owners, shares, _i, _b, owner, newReferrerTreasuryPDA, newTreasury, _c, _d, referrerTreasuryATA, _e, _f;
        return tslib_1.__generator(this, function (_g) {
            switch (_g.label) {
                case 0:
                    instructions = [];
                    return [4 /*yield*/, program.account.treasury.fetchNullable(treasuryPDA)];
                case 1:
                    treasury = _g.sent();
                    if (!treasury)
                        throw type_1.TREASURY_DOESNT_EXIST;
                    if (!treasury.owners.find(function (owner) { return owner.ownerPda.toString() === buddyPDA.toString(); }))
                        throw type_1.TREASURY_OWNER_MISMATCH;
                    return [4 /*yield*/, program.account.masterOrganization.all()];
                case 2:
                    masterOrg = (_g.sent())[0];
                    masterTreasuryPDA = (0, pda_2.getMasterTreasuryPDA)(program, treasury.mint);
                    return [4 /*yield*/, (0, pda_1.getTokenAccount)(masterTreasuryPDA, treasury.mint)];
                case 3:
                    masterTreasuryATA = _g.sent();
                    return [4 /*yield*/, program.account.buddy.fetchNullable(buddyPDA)];
                case 4:
                    buddy = _g.sent();
                    if (!buddy)
                        throw type_1.BUDDY_DOESNT_EXIST;
                    profilePDA = buddyPDA;
                    if (!!((_a = buddy.buddyType) === null || _a === void 0 ? void 0 : _a.profile)) return [3 /*break*/, 6];
                    return [4 /*yield*/, program.account.buddy.all([
                            { memcmp: { offset: 8, bytes: buddy.authority.toBase58() } },
                        ])];
                case 5:
                    profile = (_g.sent())[0];
                    profilePDA = profile.publicKey;
                    _g.label = 6;
                case 6: return [4 /*yield*/, (0, pda_1.getTokenAccount)(treasuryPDA, treasury.mint)];
                case 7:
                    buddyTreasuryATA = _g.sent();
                    return [4 /*yield*/, program.account.treasury.fetch(buddy.referrerTreasuryPda)];
                case 8:
                    referrerTreasury = _g.sent();
                    referrerTreasuryRewards = buddy.referrerTreasuryPda;
                    if (!(referrerTreasury.mint.toString() !== treasury.mint.toString())) return [3 /*break*/, 11];
                    owners = [], shares = [];
                    for (_i = 0, _b = referrerTreasury.owners; _i < _b.length; _i++) {
                        owner = _b[_i];
                        owners.push(owner.ownerPda);
                        shares.push(owner.share);
                    }
                    newReferrerTreasuryPDA = (0, pda_1.getTreasuryPDA)(program, owners, shares, (0, parse_1.parsePublicKey)(treasury.mint));
                    return [4 /*yield*/, program.account.treasury.fetchNullable(newReferrerTreasuryPDA)];
                case 9:
                    newTreasury = _g.sent();
                    referrerTreasuryRewards = newReferrerTreasuryPDA;
                    if (!!newTreasury) return [3 /*break*/, 11];
                    _d = (_c = instructions).push;
                    return [4 /*yield*/, (0, initTreasury_1.initTreasury)(program, authority, owners, shares, (0, parse_1.parsePublicKey)(treasury.mint))];
                case 10:
                    _d.apply(_c, [_g.sent()]);
                    _g.label = 11;
                case 11: return [4 /*yield*/, (0, pda_1.getTokenAccount)(referrerTreasuryRewards, treasury.mint)];
                case 12:
                    referrerTreasuryATA = _g.sent();
                    _f = (_e = instructions).push;
                    return [4 /*yield*/, program.methods
                            .transferUp()
                            .accounts({
                            associatedTokenProgram: spl_token_1.ASSOCIATED_TOKEN_PROGRAM_ID,
                            tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
                            systemProgram: anchor.web3.SystemProgram.programId,
                            rent: anchor.web3.SYSVAR_RENT_PUBKEY,
                            authority: authority,
                            masterOrganization: masterOrg.publicKey,
                            masterOrganizationTreasury: masterTreasuryPDA,
                            masterOrganizationTokenAccount: masterTreasuryATA,
                            buddyProfile: profilePDA,
                            buddy: buddyPDA,
                            buddyTreasury: treasuryPDA,
                            buddyTokenAccount: buddyTreasuryATA,
                            mint: (0, parse_1.parsePublicKey)(treasury.mint),
                            referrerTreasury: buddy.referrerTreasuryPda,
                            referrerTreasuryForReward: referrerTreasuryRewards,
                            referrerTokenAccount: referrerTreasuryATA
                        })
                            .instruction()];
                case 13:
                    _f.apply(_e, [_g.sent()]);
                    return [2 /*return*/, instructions];
            }
        });
    });
}
exports.transferUp = transferUp;
